features/include/pcl/features/impl/good.hpp

```diff
@@ -0,0 +1,666 @@
+inline std::vector<std::vector<unsigned int> > pcl::GOODEstimation<PointInT, NumberOfBins>::initializing2DHistogram (unsigned int number_of_bins)
+{
+  std::vector<unsigned int> row (number_of_bins, 0);
+  std::vector<std::vector<unsigned int> > histogram2D (number_of_bins, row);
+  return histogram2D;
+}
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename PointInT, int NumberOfBins>
+void
+pcl::GOODEstimation<PointInT, NumberOfBins>::computeFeature (PointCloudOut &output )	
+{
+  std::vector<float> object_description;
+  double largest_side = 0;
+  int  sign = 1;
+  std::vector <float> view_point_entropy;
+  std::string name_of_sorted_projected_plane;
+  PointCloudIn initial_cloud_projection_along_x_axis (new pcl::PointCloud<PointInT>);//Declare a boost share ptr to the pointCloud
+  PointCloudIn initial_cloud_projection_along_y_axis (new pcl::PointCloud<PointInT>);//Declare a boost share ptr to the pointCloud
+  PointCloudIn initial_cloud_projection_along_z_axis (new pcl::PointCloud<PointInT>);//Declare a boost share ptr to the pointCloud
+  pcl::PointXYZ pt; 
+  
+  /* __________________________
+  |                            |
+  | construct ORF based on PCA |
+  |____________________________| */
+
+  ///////// the theory of new shape descriptor ///////////////////////////
+  // //NOTE  the PCA base reference frame construction basically does:
+  // 1) compute the centroid (c0, c1, c2) and the normalized covariance
+  // 2) compute the eigenvectors e0, e1, e2. The reference system will be (e0, e1, e0 X e1) --- note: e0 X e1 = +/- e2
+  // 3) move the points in that RF --- note: the transformation given by the rotation matrix (e0, e1, e0 X e1) & (c0, c1, c2) must be inverted
+  // 4) compute the max, the min and the center of the diagonal (mean_diag)
+  // 5) given a box centered at the origin with size (max_pt.x - min_pt.x, max_pt.y - min_pt.y, max_pt.z - min_pt.z) 
+  //    the transformation you have to apply is Rotation = (e0, e1, e0 X e1) & Translation = Rotation * mean_diag + (c0, c1, c2)
+  
+  // compute principal directions	  
+  Eigen::Vector4f centroid;
+  pcl::compute3DCentroid (*input_, centroid);
+  Eigen::Matrix3f covariance;
+  computeCovarianceMatrixNormalized (*input_, centroid, covariance);
+  Eigen::SelfAdjointEigenSolver<Eigen::Matrix3f> eigen_solver (covariance, Eigen::ComputeEigenvectors);
+  Eigen::Matrix3f eigen_vectors = eigen_solver.eigenvectors();
+  eigen_vectors.col(2) = eigen_vectors.col(0).cross(eigen_vectors.col(1));
+  //std::cout << "eigen vectores :  \n " << eigen_vectors << std::endl;
+  
+  //Eigen::Vector3f eigen_values =  eigen_solver.eigenvalues();
+  //std::cout << "The eigenvalues of the covariance matrix before sorting are \n:" << eigen_values << endl;
+  
+  //sorting eigen vectors based on eigen values
+  eigen_vectors.col(0) = eigen_vectors.col(2);
+  eigen_vectors.col(2) = eigen_vectors.col(0).cross (eigen_vectors.col(1));
```

Comment:
You changed first column in the previous line, so effectively this reads as:

```cpp
eigen_vectors.col(2) = eigen_vectors.col(2).cross (eigen_vectors.col(1));
```

Are you sure this is intended?